### PR TITLE
GPU

### PR DESCRIPTION
**Pull request by barakugav**

### CODE REVIEW

**Comment by poja** [comment]
You are TOTAHH.
I have a crazy week so I'll check it out on the weekend.

**Comment by barakugav** [comment]
:)
dont merge it though

File: src/game/self_play_cmd.rs [hunk 1]
```diff
@@ -194,47 +151,81 @@ impl<Game: IGame> Builder<MCTSPlayer<Game>> for PlayerBuilder<Game> {
     }
 }
 
+struct CacheBuilder<Game: IGame> {
+    max_size: usize,
+    caches: Vec<Arc<ValueFuncCache<Game>>>,
+}
+impl<Game: IGame> CacheBuilder<Game> {
```
Comment (poja) [inline src/game/self_play_cmd.rs:158]: Confusing names here?

File: train/main.py [hunk 2]
```diff
@@ -27,5 +27,12 @@
     with open(args.config, "r") as config_file:
         config = yaml.safe_load(config_file)
 
+    if True or not config["cpu"]:
```
Comment (poja) [inline train/main.py:30]: don't forget to remove this True

**Review by poja — APPROVED** [review]
I only read briefly but it looks very cool

File: src/game/net.rs [hunk 3]
```diff
@@ -106,18 +194,120 @@ impl<Game: IGame, const CPU: bool> TwoHeadedNetBase<Game, CPU> {
...
         to_planes: &impl Fn(&Game::Position) -> Vec<Game::Bitboard>,
     ) -> (f32, Vec<(Game::Move, f32)>) {
-        let planes = to_planes(pos);
-        let input = planes_to_tensor::<Game, CPU>(planes);
-        let (val, move_scores) = self.run_net(input);
+        /* Add a new sample to the current batch */
+        let batch;
+        let batch_idx;
+        let sample_idx;
+        let mut run_net = {
```
Comment (poja) [inline src/game/net.rs:201]: Ideally the batching code shouldn't be under TwoHeadedBaseNet, but realistically it's the only relevant net so not really a problem

File: train/main.py (hunk 2 above)
Comment (barakugav) [reply train/main.py:30]: done

**Pull request merged**